File: src/envelope/export.py
"""
Proof envelope export module.

Wraps a scored local receipt into a proof envelope for submission
to proofofship.

Spec: docs/specs/local-to-global-proof-envelope-v1.md
"""

from __future__ import annotations

import json
import re
import time
import random
from datetime import datetime, timezone

from ..scoring.hash_validator import compute_content_hash


# ---------------------------------------------------------------------------
# ULID generation (no external dependency)
# ---------------------------------------------------------------------------

_CROCKFORD = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"


def _generate_ulid() -> str:
    """Generate a ULID-like string (timestamp + random)."""
    ts_ms = int(time.time() * 1000)
    ts_chars = []
    for _ in range(10):
        ts_chars.append(_CROCKFORD[ts_ms & 0x1F])
        ts_ms >>= 5
    ts_chars.reverse()
    rand_chars = [random.choice(_CROCKFORD) for _ in range(16)]
    return "".join(ts_chars) + "".join(rand_chars)


# ---------------------------------------------------------------------------
# Actor extraction
# ---------------------------------------------------------------------------

_GITHUB_URL_RE = re.compile(
    r"https?://github\.com/([a-zA-Z0-9](?:[a-zA-Z0-9]|-(?=[a-zA-Z0-9])){0,38})"
)


def extract_actor(receipt: dict) -> dict:
    """Extract actor identity from receipt subject."""
    subject = receipt.get("subject", {})
    profiles = subject.get("profiles", [])

    github_username = None
    profile_urls = []

    for p in profiles:
        url = p.get("url", "")
        if url:
            profile_urls.append(url)
        if p.get("kind") == "github":
            match = _GITHUB_URL_RE.match(url)
            if match:
                github_username = match.group(1)

    if not github_username:
        raise ValueError("No GitHub profile found in receipt subject.profiles")

    return {
        "github_username": github_username,
        "display_name": subject.get("name", github_username),
        "profile_urls": profile_urls,
    }


# ---------------------------------------------------------------------------
# Export
# ---------------------------------------------------------------------------


def export_proof_envelope(
    receipt: dict,
    game_state: dict | None = None,
) -> dict:
    """
    Convert a receipt into a proof envelope.

    Args:
        receipt: Schema-valid receipt dict.
        game_state: Optional local game state for score snapshot.

    Returns:
        Proof envelope dict conforming to proof-envelope.v1.json.

    Raises:
        ValueError: If no GitHub profile is found.
    """
    actor = extract_actor(receipt)
    content_hash = compute_content_hash(receipt)

    envelope: dict = {
        "envelope_version": "1.0",
        "envelope_id": _generate_ulid(),
        "content_hash": content_hash,
        "submitted_at": datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
        "actor": actor,
        "receipt": receipt,
        "export_metadata": {
            "generator": "ship-receipts",
            "generator_version": "0.1.0",
            "ship_receipts_schema_version": receipt.get("version", "0.1"),
        },
    }

    # Pass through receipt_id if present
    if receipt.get("receipt_id"):
        envelope["receipt_id"] = receipt["receipt_id"]

    # Pass through DR attestation if embedded in receipt
    if receipt.get("attestation"):
        envelope["attestation"] = receipt["attestation"]

    # Optional local score snapshot
    if game_state:
        snapshot = _extract_local_snapshot(receipt, game_state)
        if snapshot:
            envelope["local_score_snapshot"] = snapshot

    return envelope


def _extract_local_snapshot(receipt: dict, game_state: dict) -> dict | None:
    """Extract local score snapshot from game state history."""
    content_hash = receipt.get("meta", {}).get("content_hash")
    if not content_hash:
        content_hash = compute_content_hash(receipt)

    for entry in game_state.get("history", []):
        if entry.get("receipt_hash") == content_hash:
            return {
                "base_score": entry["breakdown"]["base"],
                "final_score": entry["score"],
                "streak_days": game_state.get("streak", {}).get("current", 0),
                "streak_multiplier": entry["breakdown"].get("streak_multiplier", 1.0),
                "integrity_multiplier": entry["breakdown"].get("integrity_multiplier", 1.0),
                "computed_at": datetime.now(timezone.utc).isoformat(),
            }

    return None

File: src-ts/envelope/export.ts
import { readFile } from "node:fs/promises";
import { join } from "node:path";
import { ulid } from "ulid";
import { computeContentHash } from "../scoring/hash-validator.js";

type Receipt = Record<string, any>;
type GameState = Record<string, any> | null | undefined;

const GITHUB_URL_RE = /^https?:\/\/github\.com\/([a-zA-Z0-9](?:[a-zA-Z0-9]|-(?=[a-zA-Z0-9])){0,38})/;

export function extractActor(receipt: Receipt): Record<string, any> {
  const subject = receipt.subject ?? {};
  const profiles = Array.isArray(subject.profiles) ? subject.profiles : [];
  let githubUsername: string | null = null;
  const profileUrls: string[] = [];

  for (const p of profiles) {
    const url = typeof p?.url === "string" ? p.url : "";
    if (url) profileUrls.push(url);
    if (p?.kind === "github") {
      const match = url.match(GITHUB_URL_RE);
      if (match?.[1]) githubUsername = match[1];
    }
  }

  if (!githubUsername) {
    throw new Error("No GitHub profile found in receipt subject.profiles");
  }

  return {
    github_username: githubUsername,
    display_name: subject.name ?? githubUsername,
    profile_urls: profileUrls,
  };
}

function extractLocalSnapshot(receipt: Receipt, gameState: Record<string, any>): Record<string, any> | null {
  const contentHash = receipt?.meta?.content_hash || computeContentHash(receipt);
  const history = Array.isArray(gameState?.history) ? gameState.history : [];
  for (const entry of history) {
    if (entry?.receipt_hash === contentHash) {
      return {
        base_score: entry.breakdown.base,
        final_score: entry.score,
        streak_days: gameState?.streak?.current ?? 0,
        streak_multiplier: entry.breakdown.streak_multiplier ?? 1.0,
        integrity_multiplier: entry.breakdown.integrity_multiplier ?? 1.0,
        computed_at: new Date().toISOString(),
      };
    }
  }
  return null;
}

export function exportProofEnvelope(receipt: Receipt, gameState?: GameState): Record<string, any> {
  const actor = extractActor(receipt);
  const contentHash = computeContentHash(receipt);

  const envelope: Record<string, any> = {
    envelope_version: "1.0",
    envelope_id: ulid(),
    content_hash: contentHash,
    submitted_at: new Date().toISOString().replace(/\.\d{3}Z$/, "Z"),
    actor,
    receipt,
    export_metadata: {
      generator: "ship-receipts",
      generator_version: "0.1.0",
      ship_receipts_schema_version: receipt.version ?? "0.1",
    },
  };

  if (gameState) {
    const snapshot = extractLocalSnapshot(receipt, gameState);
    if (snapshot) envelope.local_score_snapshot = snapshot;
  }

  return envelope;
}

export async function readGameStateIfPresent(rootDir = "."): Promise<Record<string, any> | null> {
  try {
    const raw = await readFile(join(rootDir, ".ship-receipts", "game-state.json"), "utf8");
    return JSON.parse(raw);
  } catch {
    return null;
  }
}
